04_Userspace/02_Switching_Modes.md

```diff
@@ -0,0 +1,92 @@
+# Switching Modes
+
+## Getting to User Mode
+There are a few ways to do this, but the most straight foward way is the use the `iret` instruction.
+
+`iret` pops 5 arguments off of the stack, and then performs several operations atomically:
+
+- It pops `rip` and `cs` from the stack, this is like a far jump/return. `cs` sets the mode for instruction fetches, and `rip` is the first instruction to run after `iret`.
+- It pops `rflags` into the flags register.
+- It pops `rsp` and `ss` from the stack. This changes the mode for data accesses, and the stack used after `iret`.
+
+This is a very powerful instruction because it allows us to change the mode of both code and data accesses, as well as jump to new code all at once. It has the added benefit of switching the stack and flags at the same time, which is fantastic. This is everything we need to properly jump to user code.
+
+Changing the flags atomically like this means we can go from having interrupts disabled in supervisor mode, to interrupts enabled in user code. All without the risk of having an interrupt occuring while we change these values ourselves.
+
+### What to Push Onto The Stack
+Now let's talk about what these values should be. `rflags` is an easy one: set it to 0x202. Bit 2 is a legacy feature and must always be set, the other bit (`0x200`) is the interrupt enable flag. This means all other flags are cleared, and is what C/C++ and other languages expect flags to look like when starting a program.
+
+For `ss` and `cs` it depends on the layout of your GDT. We'll assume that you have 5 entries in your GDT:
+
+- 0x00, Null
+- 0x08, Supervisor Code (ring 0)
+- 0x10, Supervisor Data (ring 0)
+- 0x18, User Code (ring 3)
+- 0x20, User Data (ring 3)
+
+Now `ss` and `cs` are *selectors*, which you'll remember is not just the byte offset into the gdt, but the lower two bits contain a field called RPL. Requested Priviledge Level is a legacy feature, but it's still enforced by the cpu, so we have to use it. RPL is a sort of 'override' for the target ring, it's useful in some edge cases, but otherwise is best set to the ring you want to jump to.
```

Comment:
Interesting! To me this seems correct in english, switching it to plural also seems fine:
"which you'll remember are not just the byte offset into the gdt"

Reply:
ah ok, if i remeber well  in italian if you start a sentence with singular and refer to the same thing you ten to keep it singular till the end. 
but if it is correct, well is fine even because i can't say that i have a wide knowledge of english grammar XD

Reply:
I changed it anyway :)